portable/GCC/ARM_CRx_MPU/portASM.S

```diff
@@ -0,0 +1,624 @@
+*/
+.align 4
+.global prvMpuSetRegion
+.type prvMpuSetRegion, %function
+prvMpuSetRegion:
+    /* Only 15 possible regions, drop all other bits */
+    AND     R0,  R0, #15
+    /* Select the MPU Region selected by ulRegionNumber */
+    MCR     p15, #0, R0, c6, c2, #0
+    /* Set the Base Address to be ulBaseAddress */
+    MCR     p15, #0, R1, c6, c1, #0
+    /* Set the Access Attributes to be ulRegionPermissions */
+    MCR     p15, #0, R3, c6, c1, #4
+    /* Set the Size and Enable bits to be ulRegionSize */
+    MCR     p15, #0, R2, c6, c1, #2
+    /* Return to caller */
+    BX      LR
+
+/* ----------------------------------------------------------------------------------- */
+/* Set the Enable bit of the MPU Enable Register to 1. */
+.align 4
+.global prvMpuEnable
+.type prvMpuEnable, %function
+prvMpuEnable:
+    /* Read the current MPU control register into R0 */
+    MRC     p15, #0, R0, c1, c0, #0
+    /* Set the enable bit to high */
+    ORR     R0,  R0, #0x1
+    /* Data sync */
+    DSB
+    /* Write out previous MPU control register with a high enable bit */
+    MCR     p15, #0, R0, c1, c0, #0
+    /* Instruction sync */
+    ISB
+    /* Return to caller */
+    BX      LR
+
+/* ----------------------------------------------------------------------------------- */
+/* Set the Enable bit of the MPU Enable Register to 0. */
+.align 4
+.global prvMpuDisable
+.type prvMpuDisable, %function
+prvMpuDisable:
+    /* Read the MPU enable register values into R0 */
+    MRC     p15, #0, R0, c1, c0, #0
+    /* Perform a bitwise AND of R0 and NOT #1, i.e. clear bit 1 */
+    BIC     R0,  R0, #1
+    /* Wait for all pending explicit data accesses to complete */
+    DSB
+    /* Write out to the MPU Enable Register */
+    MCR     p15, #0, R0, c1, c0, #0
+    /* Flushes the pipeline and prefetch buffer(s) in the processor. */
+    /* Ensures all following instructions are fetched from cache or memory. */
+    ISB
+    /* Return to caller */
+    BX      LR
+
+.align 4
+.global FreeRTOS_IRQ_Handler
+.type FreeRTOS_IRQ_Handler, %function
+FreeRTOS_IRQ_Handler:
+    /* Return to the interrupted instruction. */
+    SUB     LR, LR, #4
+
+    /* Save the return state to the IRQ stack */
+    SRS     SP!, #IRQ_MODE
+
+    /* Push used registers. */
+    PUSH    {R0-R3, R12}
+
+    /* Increment nesting count. R3 holds the address of ulPortInterruptNesting
+    for future use. R1 holds the original ulPortInterruptNesting value for
+    future use. */
+    LDR     R3, =ulPortInterruptNesting
+    LDR     R1, [R3]
+    ADD     R0, R1, #1
+    STR     R0, [R3]
+
+    /* Ensure bit 2 of the stack pointer is clear. R2 holds the bit 2 value for
+    future use. */
+    MOV     R0, SP
+    AND     R2, R0, #4
+    SUB     SP, SP, R2
+
+    /* Call the interrupt handler. */
+    PUSH    {R0-R3, LR}
+    BL      vApplicationIRQHandler
+
+    /* Restore the registers */
+    POP     {R0-R3, LR}
+    ADD     SP, SP, R2
+
+    CPSID    i
```

Comment:
Nit: Does this need to be capital `I`? It has been in the other disable/enable calls

Reply:
I don't think so, but thanks for pointing this out. I intentionally was trying to use capitals for ASM work.
Can you please add it as a suggestion to ensure this gets fixed?

Reply:
Changed this to a capital